File: aoc_2022_javascript/day20.js
class Node {
  constructor(data, next = null, prev = null) {
    this.data = data;
    this.next = next;
    this.prev = prev;
  }
}

class CircularLinkedList {
  constructor() {
    this.first = null;
    this.list = [];
  }
  
  insertFirst(data) {
    const node = new Node(data, this.first);
    this.first = node;
  }

  insertAfter(refNode, newNode) {
    newNode.prev = refNode;
    newNode.next = refNode.next;
    newNode.next.prev = newNode;
    refNode.next = newNode;
  }

  insertAtEnd(newNode) {
    if (this.first == null) {
      this.first = newNode;
      newNode.next = newNode;
      newNode.prev = newNode;
    } else {
      this.insertAfter(this.first.prev, newNode);
    }
    this.list.push(newNode);
  }

  remove(node) {
    if (this.first.next == this.first) {
      this.first = null;
    } else {
      node.prev.next = node.next;
      node.next.prev = node.prev;
      if (this.first == node) {
        this.first = node.next;
      }
    }
  }

  print() {
    let node = this.first;
    for (let i = 0; i < this.list.length; i++) {
      console.log(node.data);
      node = node.next;
    }
  }
}

main();

function main() {
  const fs = require("fs");
  const input = fs
    .readFileSync("input/day20.txt", { encoding: "utf-8" })
    .trim()
    .split("\r\n");
  let start = Date.now();
  solution = solve(input);
  let end = Date.now();
  console.log(`Execution time: ${end - start} ms`);
  console.log(`Part 1: ${solution[0]}`);
  console.log(`Part 2: ${solution[1]}`);
}

function solve(lines) {
  let part1 = decrypt(lines, true);
  let part2 = decrypt(lines, false);
  
  return [part1, part2];
}

function decrypt(lines, firstPart) {
  let list = new CircularLinkedList();
  let nodeZero = null;
  let decryptionKey = 811589153;
  if (firstPart) {
    decryptionKey = 1;
  }
  for (let line of lines) {
    let n = new Node((parseInt(line) * decryptionKey));
    list.insertAtEnd(n);
    if (n.data == 0) {
      nodeZero = n;
    }
  }
  let rounds = 10;
  if (firstPart) {
    rounds = 1;
  }
  for (let a = 0; a < rounds; a++) {
    for (let node of list.list) {
      let targetNode = node.prev;
      list.remove(node);
      for (let i = 0; i < Math.abs(node.data) % (list.list.length - 1); i++) {
        if (node.data < 0) {
          targetNode = targetNode.prev;
        } else if (node.data > 0) {
          targetNode = targetNode.next;
        }
      }
      list.insertAfter(targetNode, node);
    }
  }
  
  let ret = 0;
  let node = nodeZero;
  for (let i = 0; i < 3001; i++) {
    if (i == 1000 || i == 2000 || i == 3000) {
      ret += node.data;
    }
    node = node.next;
  }
  return ret;
}

File: aoc_2024_python/day20.py
import aoc

def main(lines: list):
    for y, line in enumerate(lines):
        for x, c in enumerate(line):
            if c == "S":
                start = (x, y)
            if c == "E":
                end = (x, y)
    
    path = []
    pos = start
    while True:
        path.append(pos)
        if pos == end:
            break
        for dir in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
            neighbor = (pos[0] + dir[0], pos[1] + dir[1])
            if lines[neighbor[1]][neighbor[0]] != "#" and (len(path) < 2 or neighbor != path[-2]):
                pos = neighbor
                break
    
    part1 = 0
    part2 = 0
    for a, start in enumerate(path):
        for b in range(a + 101, len(path)):
            end = path[b]
            cheat_distance = abs(start[0] - end[0]) + abs(start[1] - end[1])
            saved_distance = b - a - cheat_distance
            if cheat_distance <= 2 and saved_distance >= 100:
                part1 += 1
            if cheat_distance <= 20 and saved_distance >= 100:
                part2 += 1

    return part1, part2

aoc.run_lines(main, "day20.txt")
